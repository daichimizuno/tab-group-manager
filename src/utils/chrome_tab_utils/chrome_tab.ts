import { TabGroup } from "../../dao/chrome_storage_access";

export default class ChromeTabAccess {
  createTabGroup = (tabGroup: TabGroup[]) => {
    chrome.runtime.sendMessage(tabGroup);
  };
}
