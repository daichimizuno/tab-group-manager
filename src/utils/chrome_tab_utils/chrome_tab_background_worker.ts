import { Color, TabGroup } from "../../dao/chrome_storage_access";

export const getAllInWindow = async (): Promise<chrome.tabs.Tab[]> => {
  const tabs = await chrome.tabs.query({ currentWindow: true });
  return tabs;
};

const openTab = (color: Color, title: string, urls: string[]) => {
  // groupIdを取得するために、１つ目のurlだけopenさせる
  chrome.tabs.create({ url: urls[0] }, (tab) => {
    chrome.tabs.group({ tabIds: tab.id }, (gid: number) => {
      chrome.tabGroups.update(
        gid,
        { color: color, title: title },
        (group) => {}
      );

      // 残りのurlを取得したopenして、groupIdに紐付けさせる
      // グループの色、タイトルはupdateで完了しているので、ここでは行わない
      urls.forEach((url, index) => {
        if (index !== 0) {
          chrome.tabs.create({ url: url }, (tab) => {
            chrome.tabs.group(
              { tabIds: tab.id, groupId: gid },
              (gid: number) => {}
            );
          });
        }
      });
    });
  });
};

chrome.runtime.onMessage.addListener((tabGroup: TabGroup[]) => {
  tabGroup.forEach((tab) => {
    openTab(tab.tabColor, tab.tabGroupName, tab.urls);
  });
  return Promise.resolve();
});
