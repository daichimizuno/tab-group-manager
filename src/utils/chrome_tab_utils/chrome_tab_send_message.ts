import { TabGroup } from "../../dao/chrome_storage_access";

export default class ChromeTabSendMessage {
  createTabGroupsSendMessage = (tabGroup: TabGroup[]) => {
    chrome.runtime.sendMessage(tabGroup);
  };
}
