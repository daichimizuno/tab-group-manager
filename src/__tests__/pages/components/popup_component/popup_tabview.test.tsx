import {
  fireEvent,
  render,
  RenderResult,
  screen,
} from "@testing-library/react";
import React from "react";
import renderer, { act, ReactTestRendererJSON } from "react-test-renderer";
import PopupTabView from "../../../../pages/components/popup_component/popup_tabview";
import * as ChromeTabBackgroundWorker from "../../../../utils/chrome_tab_utils/chrome_tab_background_worker";

const tabs: chrome.tabs.Tab[] = [
  {
    index: 1,
    title: "test",
    url: "https://google.com",
  } as chrome.tabs.Tab,
  {
    index: 2,
    title: "test2",
    url: "https://google.com",
  } as chrome.tabs.Tab,
];

let component: RenderResult;

beforeEach(() => {
  jest.resetAllMocks();
  const getAllWindowSpy = jest
    .spyOn(ChromeTabBackgroundWorker, "getAllInWindow")
    .mockResolvedValue(tabs);

  act(() => {
    component = render(<PopupTabView />);
  });
});

afterEach(() => {
  jest.resetAllMocks();
});

describe("PopupTabViewの試験", () => {
  describe("コンポーネントテスト", () => {
    describe("正常系", () => {
      test("スナップショットが前回の結果と一致しているかどうか", () => {
        let tree = renderer
          .create(<PopupTabView />)
          .toJSON() as ReactTestRendererJSON;
        expect(tree).toMatchSnapshot();
      });

      test("タブが２つ存在し、「開いているタブ」と「保存されている設定」があること", () => {
        const openedTab = screen.getByRole("tab", { name: "開いているタブ" });
        const setTab = screen.getByRole("tab", { name: "保存されている設定" });

        expect(openedTab.textContent).toBe("開いているタブ");
        expect(setTab.textContent).toBe("保存されている設定");
      });

      test("初回は「開かれているタブ」が選択されていて、「test」と「test2」というタブが表示されていること", () => {
        const tabPanel = screen.getByRole("tabpanel");
        expect(screen.getByText("test")).toBeVisible();
        expect(screen.getByText("test2")).toBeVisible();
      });

      test("「保存されている設定」を押すと「setting page」というページが開かれること", () => {
        const setTab = screen.getByRole("tab", { name: "保存されている設定" });
        act(() => {
          fireEvent.click(setTab);
        });

        const tabPanel = screen.getByRole("tabpanel");
        expect(tabPanel.textContent).toBe("setting page");
      });
    });
  });
});
