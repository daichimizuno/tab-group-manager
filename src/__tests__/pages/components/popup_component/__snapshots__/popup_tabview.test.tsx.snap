// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopupTabViewの試験 コンポーネントテスト 正常系 スナップショットが前回の結果と一致しているかどうか 1`] = `
<div
  className="MuiBox-root css-a4ih50"
>
  <div
    className="MuiBox-root css-1gsv261"
  >
    <div
      className="MuiTabs-root css-1ujnqem-MuiTabs-root"
      data-testid="tab-list"
    >
      <div
        className="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
        onScroll={[Function]}
        style={
          Object {
            "marginBottom": -0,
            "overflow": "hidden",
          }
        }
      >
        <div
          aria-label="tab-list"
          aria-orientation={null}
          className="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
          onKeyDown={[Function]}
          role="tablist"
        >
          <button
            aria-controls={null}
            aria-selected={true}
            className="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit MuiTab-fullWidth Mui-selected css-1vw5i9y-MuiButtonBase-root-MuiTab-root"
            disabled={false}
            id={null}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tab"
            tabIndex={0}
            type="button"
          >
            開いているタブ
            <span
              className="MuiTabs-indicator css-11yukd5-MuiTabs-indicator"
              style={Object {}}
            />
          </button>
          <button
            aria-controls={null}
            aria-selected={false}
            className="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit MuiTab-fullWidth css-1vw5i9y-MuiButtonBase-root-MuiTab-root"
            disabled={false}
            id={null}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tab"
            tabIndex={-1}
            type="button"
          >
            保存されている設定
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    aria-labelledby={null}
    className="MuiTabPanel-root css-13xfq8m-MuiTabPanel-root"
    hidden={false}
    id={null}
    role="tabpanel"
  />
  <div
    aria-labelledby={null}
    className="MuiTabPanel-root css-13xfq8m-MuiTabPanel-root"
    hidden={true}
    id={null}
    role="tabpanel"
  />
</div>
`;
